python/cucim/src/cucim/core/operations/color/stain_normalizer.py

```diff
@@ -0,0 +1,821 @@
+
+    # convert to absorbance (optical density) matrix of shape (3, n_pixels)
+    absorbance = _image_to_absorbance_matrix(
+        image,
+        source_intensity=source_intensity,
+        image_type=image_type,
+        channel_axis=channel_axis,
+    )
+
+    # remove transparent pixels
+    absorbance = absorbance[:, cp.all(absorbance > beta, axis=0)]
+    if absorbance.size == 0:
+        raise ValueError(
+            "All pixels of the input image are below the threshold."
+        )
+
+    # compute eigenvectors (do small 3x3 matrix calculations on the host)
+    cov = _covariance(absorbance, rowvar=True)
+    cov = cp.asnumpy(cov).astype(np.float32, copy=False)
+    _, ev = np.linalg.eigh(cov)
+    ev = ev[:, [2, 1]]
+    # flip to ensure positive first coordinate so arctan2 angles are about 0
+    if ev[0, 0] < 0:
+        ev[:, 0] *= -1
+    if ev[0, 1] < 0:
+        ev[:, 1] *= -1
+
+    # project on the plane spanned by the eigenvectors
+    projection = cp.dot(cp.asarray(ev.T), absorbance)
+
+    # find the vectors that span the whole data (min and max angles)
+    phi = cp.arctan2(projection[1], projection[0])
+    min_phi, max_phi = cp.percentile(phi, (alpha, 100 - alpha))
+    # need these scalars on the host
+    min_phi, max_phi = float(min_phi), float(max_phi)
+
+    # project back to absorbance space
+    v_min = np.array([math.cos(min_phi), math.sin(min_phi)], dtype=np.float32)
+    v_max = np.array([math.cos(max_phi), math.sin(max_phi)], dtype=np.float32)
+    v1 = np.dot(ev, v_min)
+    v2 = np.dot(ev, v_max)
+
+    # Make Hematoxylin (H) first and eosin (E) second by comparing the
+    # R channel value
+    if v1[0] < v2[0]:
+        v1, v2 = v2, v1
+    stain_coeff = np.stack((v1, v2), axis=-1)
+
+    # renormalize columns to reduce numerical error
+    stain_coeff /= np.linalg.norm(stain_coeff, axis=0, keepdims=True)
+    if append_third_column:
+        stain_coeff = _complement_stain_matrix(stain_coeff)
+    return cp.asarray(stain_coeff)
+
+
+def _get_raw_concentrations(src_stain_coeff, absorbance, method):
+
+    if absorbance.ndim != 2 or absorbance.shape[0] != 3:
+        raise ValueError("`absorbance` must be shape (3, n_pixels)")
+
+    # estimate the raw stain concentrations
+    if method == 'lstsq':
+        if src_stain_coeff.shape != (3, 2):
+            raise ValueError("expected a set of two stain vectors")
+        # If there are only two stain vectors, use a least-squares estimate.
+        conc_raw = cp.linalg.lstsq(
+            src_stain_coeff, absorbance, rcond=None
+        )[0]
+    elif method == 'ortho':
+        if src_stain_coeff.shape != (3, 3):
+            raise ValueError("expected a set of three stain vectors")
+        # This approach relies on a square stain coeffs matrix as used by
+        # HistomicsTK. In practice, it gives nearly identical results to the
+        # least-squares approach.
+        coeff_inv = cp.linalg.inv(src_stain_coeff)
```

Comment:
Instead of creating the extra column for statin coefficients, can't we use pseudo-inverse (for rectangular matrices)? `{X^T.X}^{-1}X^T`

Reply:
Isn't this pinv-based solution just "least squares"? Perhaps we should just remove the `method` argument for now and always use the psuedo-inverse code path.

I was thinking `method` could be useful later if we were to add additional methods like LASSO or a non-negative matrix factorization that would ensure non-negative concentrations.

Reply:
I like that you added different methods but if the gain is not that much for now, we can keep only one of the methods and add the argument if needed in the future.